patches/enable-instcombine-switch.patch

```diff
@@ -0,0 +1,367 @@
+ }
+ 
+@@ -502,7 +509,8 @@
+ 
+   MPM.add(createDeadArgEliminationPass()); // Dead argument elimination
+ 
+-  addInstructionCombiningPass(MPM); // Clean up after IPCP & DAE
++  if (!DisableInstCombine)
++    addInstructionCombiningPass(MPM); // Clean up after IPCP & DAE
+   addExtensionsToPM(EP_Peephole, MPM);
+   MPM.add(createCFGSimplificationPass()); // Clean up after IPCP & DAE
+ 
+@@ -638,7 +646,8 @@
+   // on -O1 and no #pragma is found). Would be good to have these two passes
+   // as function calls, so that we can only pass them when the vectorizer
+   // changed the code.
+-  addInstructionCombiningPass(MPM);
++  if (!DisableInstCombine)
++    addInstructionCombiningPass(MPM);
+   if (OptLevel > 1 && ExtraVectorizerPasses) {
+     // At higher optimization levels, try to clean up any runtime overlap and
+     // alignment checks inserted by the vectorizer. We want to track correllated
+@@ -648,11 +657,13 @@
+     // dead (or speculatable) control flows or more combining opportunities.
+     MPM.add(createEarlyCSEPass());
+     MPM.add(createCorrelatedValuePropagationPass());
+-    addInstructionCombiningPass(MPM);
++    if (!DisableInstCombine)
++      addInstructionCombiningPass(MPM);
+     MPM.add(createLICMPass());
+     MPM.add(createLoopUnswitchPass(SizeLevel || OptLevel < 3, DivergentTarget));
+     MPM.add(createCFGSimplificationPass());
+-    addInstructionCombiningPass(MPM);
++    if (!DisableInstCombine)
++      addInstructionCombiningPass(MPM);
+   }
+ 
+   // Cleanup after loop vectorization, etc. Simplification passes like CVP and
+@@ -670,7 +681,8 @@
+   }
+ 
+   addExtensionsToPM(EP_Peephole, MPM);
+-  addInstructionCombiningPass(MPM);
++  if (!DisableInstCombine)
++    addInstructionCombiningPass(MPM);
+ 
+   if (!DisableUnrollLoops) {
+     if (EnableUnrollAndJam) {
+@@ -683,7 +695,8 @@
+     MPM.add(createLoopUnrollPass(OptLevel));    // Unroll small loops
+ 
+     // LoopUnroll may generate some redundency to cleanup.
+-    addInstructionCombiningPass(MPM);
++    if (!DisableInstCombine)
++      addInstructionCombiningPass(MPM);
+ 
+     // Runtime unrolling will introduce runtime check in loop prologue. If the
+     // unrolled loop is a inner loop, then the prologue will be inside the
+@@ -801,9 +814,11 @@
+   // simplification opportunities, and both can propagate functions through
+   // function pointers.  When this happens, we often have to resolve varargs
+   // calls, etc, so let instcombine do this.
+-  if (OptLevel > 2)
+-    PM.add(createAggressiveInstCombinerPass());
+-  addInstructionCombiningPass(PM);
++  if (!DisableInstCombine) {
++    if (OptLevel > 2)
++      PM.add(createAggressiveInstCombinerPass());
++    addInstructionCombiningPass(PM);
++  }
+   addExtensionsToPM(EP_Peephole, PM);
+ 
+   // Inline small functions
+@@ -825,7 +840,8 @@
+   PM.add(createArgumentPromotionPass());
+ 
+   // The IPO passes may leave cruft around.  Clean up after them.
+-  addInstructionCombiningPass(PM);
```

Comment:
this approach is fine if you want to take it, but you can have the same effect with less code if you just (conditionally) return early from the instcombine code. in that case the pass runs but it doesn't do anything.